aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -279,35 +282,37 @@ module aptos_framework::account {
             abort error::invalid_argument(EINVALID_SCHEME)
         };
 
+        // Construct a valid `RotationProofChallenge` that `cap_rotate_key` and `cap_update_table` will validate against.
         let curr_auth_key = from_bcs::to_address(account_resource.authentication_key);
-        // construct a RotationProofChallenge to prove that the user intends to do a key rotation
         let challenge = RotationProofChallenge {
             sequence_number: account_resource.sequence_number,
             originator: addr,
             current_auth_key: curr_auth_key,
             new_public_key: to_public_key_bytes,
         };
 
-        // verify that the challenge signed by the current private key and the previous private key are both valid
-        let curr_auth_key = verify_key_rotation_signature_and_get_auth_key(from_scheme, from_public_key_bytes, cap_rotate_key, &challenge);
-        let new_auth_key = verify_key_rotation_signature_and_get_auth_key(to_scheme, to_public_key_bytes, cap_update_table, &challenge);
+        // Assert the challenge signed by the current key is valid.
+        assert_valid_signature_and_get_auth_key(from_scheme, from_public_key_bytes, cap_rotate_key, &challenge);
+        // Assert the challenge signed by the new key is valid, and get the new authentication key.
+        let new_auth_key = assert_valid_signature_and_get_auth_key(to_scheme, to_public_key_bytes, cap_update_table, &challenge);
 
-        // update the address_map table, so that we can reference to the originating address using the current address
+        // Update the `OriginatingAddress` table, so that we can find the originating address using the latest address
+        // in the event of key recovery
         let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
-        let curr_address = from_bcs::to_address(curr_auth_key);
         let new_address = from_bcs::to_address(new_auth_key);
```

Comment:
cur_address/new_address/cur_auth_key/new_auth_key always confused me when I was reading this file.
Could you unify the naming like `xxx_auth_key_raw: vector<u8>` and `xxx_auth_key: Address` and abandon the `_address` as name for auth_keys?

Reply:
There is a larger problem here: we are often abusing the type system by declaring authentication key variables using the `address` type. Technically, an `address` type is, well, for an address! And an authentication key is NOT (necessarily) an address.

But because it is sometimes useful to operate on the auth key as an `address` type, we keep abusing this...

Unfortunately, it does create confusion and we could fix it, should we decided it is important enough, which I think it is: **we do not want any confusion in our authkey/address logic which deals with key rotation, signer delegation, etc.**

I propose we create a dedicated type for authentication keys and update our API (as much as we can, in a backwards compatible manner).

Reply:
yeah I think we can continue this conversation on slack and can add an authentication key type in a different PR. 

@lightmark your point is solid, I was pretty confused about auth key vs. address too when I first started working on this. I will update the comments to make it less confusing.

Reply:
Yep, I realize :'-( We could still change everything we *can* change in a backwards-compatible manner...

Do we even have many APIs that expose the authentication key type publicly? I think it's mostly an internally-used type.

Reply:
yeah i think it's mostly internal, we do have one public API
`public fun get_authentication_key(addr: address): vector<u8> acquires Account `

could you give an example of how to change in a backwards compatible manner? I think if we're changing the type, then it'll probably breaks `struct Account`?

Reply:
We would add a `get_authentication_key_v2` and have this one abort?

Reply:
yeah i feel like that'll mess up previous callers... say if they have a struct and auth key is one of the field, they'll suddenly have a lot of aborts, and to be compatible with this version they'll need to make a v2 of their struct too : /

either way we can prob discuss this in person some time with the rest of smart contract folks and see if we can come up with a good way of handling this